services/matching.js

```diff
@@ -0,0 +1,171 @@
+   * all the properties that depend on the order of coordinates into
+   * one object for ease of use.
+   *
+   * @typedef {Object} MatchPath
+   * @property {Array<number>} coordinates - An array containing (longitude, latitude).
+   * @property {'unrestricted'|'curb'} [approach="unrestricted"] - Used to indicate how requested routes consider from which side of the road to approach a waypoint.
+   * @property {number} [radius=5] - A number in meters indicating the assumed precision of the used tracking device.
+   * @property {boolean} [isWaypoint=true] - Whether this coordinate is waypoint or not. Note! the first and last coordinates will always have to be true.
+   * @property {string} [waypointName] - Custom names for waypoint used for the arrival instruction in banners and voice instructions.
+   * @property {date} [timestamp] - Unix timestamp corresponding the coordinate.
+   */
+  config.matchPath.forEach(function(obj) {
+    matchPath.coordinates.push(obj.coordinates[0] + ',' + obj.coordinates[1]);
+
+    // isWaypoint
+    if (obj.hasOwnProperty('isWaypoint') && obj.isWaypoint != null) {
+      matchPath.isWaypoint.push(obj.isWaypoint);
+    } else {
+      matchPath.isWaypoint.push(true); // default value
+    }
+
+    ['approach', 'radius', 'waypointName', 'timestamp'].forEach(function(prop) {
+      if (obj.hasOwnProperty(prop) && obj[prop] != null) {
+        matchPath[prop].push(obj[prop]);
+      } else {
+        matchPath[prop].push('');
+      }
+    });
+  });
+
+  ['coordinates', 'approach', 'radius', 'waypointName', 'timestamp'].forEach(
+    function(prop) {
+      // avoid sending params which are all `;`
+      if (
+        matchPath[prop].every(function(value) {
+          return value === '';
+        })
+      ) {
+        delete matchPath[prop];
+      } else {
+        matchPath[prop] = matchPath[prop].join(';');
+      }
+    }
+  );
+
+  // the api requires the first and last items to be true.
+  matchPath.isWaypoint[0] = true;
+  matchPath.isWaypoint[matchPath.isWaypoint.length - 1] = true;
+
+  if (
+    matchPath.isWaypoint.every(function(value) {
+      return value === true;
+    })
+  ) {
+    delete matchPath.isWaypoint;
+  } else {
+    // the api requires the indexes to be sent
+    matchPath.isWaypoint = matchPath.isWaypoint
+      .map(function(val, i) {
+        return val === true ? i : '';
+      })
+      .join(';');
+  }
+
+  var body = pick(
+    {
+      annotations: config.annotations,
+      geometries: config.geometries,
+      language: config.language,
+      overview: config.overview,
+      steps: config.steps,
+      tidy: config.tidy,
+      approaches: matchPath.approach,
+      radiuses: matchPath.radius,
+      waypoints: matchPath.isWaypoint,
+      timestamps: matchPath.timestamp,
+      waypoint_names: matchPath.waypointName,
+      coordinates: matchPath.coordinates
+    },
+    function(_, val) {
+      return val != null;
+    }
```

Comment:
Seems like we're be using this pattern enough that it would be worth having a dedicated shortcut function, `cleanObject`? That way we can be sure we have the same implementation in each place (e.g. removing *both* `undefined` and `null` but not other falsy values).